### ISSUE TITLE
Status of packaging tools

### ISSUE THREAD

**Issue by rgommers** [Quansight-Labs/free-threaded-compatibility#7]
- [x] `packaging`: 24.0 has full support
- [ ] Build frontends:
    - [ ] `pip`: https://github.com/pypa/pip/pull/12300
    - [ ] `build`: probably nothing to do, since it depends on latest `packaging` (verify this)
- [ ] Build backends:
    - [ ] `meson-python`: https://github.com/mesonbuild/meson-python/issues/499
    - [ ] `scikit-build-core`: no issue yet
    - [ ] `setuptools`: no issue yet
- [ ] `manylinux`: https://github.com/pypa/manylinux/pull/1564 (waiting on `pip`)
- [ ] `cibuildwheel`: https://github.com/pypa/cibuildwheel/issues/1657
- [ ] wheel publishing tools (`auditwheel`, `delvewheel`, `delocate`, `repair-wheel`): to verify

**Comment by henryiii** [comment]
* Scikit-build-core done in https://github.com/scikit-build/scikit-build-core/pull/741
* manylinux done
* Cibuildwheel done except no macOS yet, waiting on beta 2 for official installers

**Comment by rgommers** [comment]
Thanks @henryiii! Table updated.

**Comment by henryiii** [comment]
Looks like both meson and CMake don't support Windows + free-threading yet, they don't know to look for the library with the "t" for linking. I think I can fix this in scikit-build-core if I can tell it the correct library from sysconfig info.

**Comment by rgommers** [comment]
Yeah, working on the same thing for Meson. I think adding the `t` in `python313t.lib` based on:
```python
bool(sysconfig.get_config_var('Py_GIL_DISABLED'))
```
should do the job. Plus the debug flavor of it (there's no limited API variant).

**Comment by henryiii** [comment]
Windows free-threaded fixed in scikit-build-core 0.9.5.

**Comment by henryiii** [comment]
FYI, sysconfig's LIBRARY is correct as far as the "t" goes but is .dll instead of .lib.

**Comment by henryiii** [comment]
Free-threading including macOS with 3.13.0b2 fully supported in cibuildwheel 2.19.0.

**Comment by rgommers** [comment]
The final release of Pip 24.1 is out 🎉. That was the biggest missing piece here, we're in a pretty happy place now.

**Comment by rgommers** [comment]
The Windows support in Meson by fixed in https://github.com/mesonbuild/meson/pull/13338. @henryiii there is a second issue on Windows beyond `libpython`, namely that `pyconfig.h` is missing the `Py_GIL_DISABLED` define (discussed in https://discuss.python.org/t/windows-installer-freethreading-and-building-extension-modules/54391/2). I only lightly browsed your CMake/scikit-build-core PRs, so maybe I missed it - but it wasn't clear to me that that fix was included.

**Comment by minrk** [comment]
Is it appropriate to add tracking of pyO3 and cffi here in build backends?

- pyO3 tracking issue: https://github.com/PyO3/pyo3/issues/4265
- cffi: https://github.com/python-cffi/cffi/issues/40

I was testing Jupyter today, and of the extension dependencies for JupyterLab, these appear to both be limiting factors. Most extensions are working fine, though:

- pyzmq works without modification when built with the cython prerelease
- argon-cffi-bindings uses cffi which unconditionally sets the unsupported py_limited_api=True (I got this to build with https://github.com/python-cffi/cffi/pull/104)
- jsonschema uses rpds-py, which uses pyO3. rpds builds, but crashes at import

so I can run a nogil _kernel_, but not the UI yet.

**Comment by rgommers** [comment]
Hey @minrk, thanks for the input! I'd say that these two aren't packaging tools so this isn't quite the right issue - and they're both important enough to deserve their own tracking issue.

PyO3 working will take a while, work has just started and it's quite complex. @ngoldbaum perhaps you could open a separate tracking issue for this?

`cffi` hasn't moved since December I think (aside from the PR you just opened), and it's the first time it has come up here. Maybe you want to open a new issue for it on this repo?

**Comment by ngoldbaum** [comment]
I opened https://github.com/Quansight-Labs/free-threaded-compatibility/issues/53 for PyO3, thanks for the reminder.

**Comment by minrk** [comment]
Thanks, I somehow overlooked that pybind11 is not in this list, which is what I was thinking of as a peer for these. Opened #54 for cffi.

**Comment by henryiii** [comment]
> but it wasn't clear to me that that fix was included.

FYI, this was fixed in CMake 3.30.3 a month or two ago.

**Comment by rgommers** [comment]
Great, thanks @henryiii!

**Comment by rgommers** [comment]
Windows wheels are also starting to materialize and that all looks good, so I ticked off `delvewheel`. I'm sure there are some corner case bugs lurking somewhere, but overall we're in good shape here.
